File: blargbot/leaderboard/ldmodule.rb
{if;!=;1;{get;~exec};
	:x: Do `{prefix}t raw leaderboard` for help!
	{return}
}
{if;==;;{get;~credit_name};
	:x: Do `{prefix}t raw leaderboard` for help! {throw;No credit name}
	{return}
}
{if;==;;{get;~leaderboard};
	:x: Do `{prefix}t raw leaderboard` for help! {throw;No leaderboard name}
	{return}
}
{//;Sets the name by uppercasing the first letter}
{set;~name;{upper;{substring;{get;~credit_name};0;1}}{substring;{get;~credit_name};1}}

{//;Checks if the array exists}
{if;!=;true;
	{isarray;{get;{get;~leaderboard}}};
	{set;{get;~leaderboard};[]}
}
{void;
	{//;Updating the leaderboard. Checks if your credits had changed in value.}
	{exec;ldupdate}
	
	{//;Getting the leaderboard}
	
	{//;The number of entries in the leaderboard}
	{set;~mx;{length;{get;{get;~leaderboard}}}}
	
	{//;The max amount of pages in the leaderboard}
	{set;~maxpg;{ceil;{math;/;{get;~mx};10}}}
	
	{//;The page that will be taken based on arguments}
	{set;~pg;
		{if;==;0;{argslength};1;
		{if;==;NaN;{parseint;{args;0}};1;
		{if;>;{args;0};{get;~maxpg};{get;~maxpg};
		{if;==;0;{floor;{args;0}};1;
		{abs;{floor;{args;0}}}
	}}}}}
	
	{//;Multiplies the page number by 10, this is the index that we will get from leaderboard}
	{set;~mg;{math;*;10;{math;-;{get;~pg};1}}}
	
	{//;Sets maximum repeats if items exceed leaderboard length}
	{if;==;{get;~maxpg};{get;~pg};
		{set;~pgr;{math;-;{get;~mx};{get;~mg}}};
		{set;~pgr;10}
	}
	
	{//;Will get rank based on the index of your userid in the leaderboard}
	{set;~rank;
		{indexof;
			{regexreplace;{get;{get;~leaderboard}};/(?:(?:\w|`)+#)(\d{1,18})/g;$1};
			{userid}
		}
	}
	
	{//;The number of zeroes preceding the numbers when showing the leaderboard. For exmaple page 10 (91-100) has [091] - [100].}
	{set;~pad;{length;{math;+;{get;~mg};10}}}
	
	{//;Makes another array including only the userids}
	{set;~0;
		{slice;
			{regexreplace;
				{get;{get;~leaderboard}};
				/(?:\d*#)(\d{1,18})/g;$1
			};
			{get;~mg};
			{math;+;0{get;~mg};10}
		}
	}
	
	{//;Makes another array for the obsolete entries}
	{if;==;1;{get;~del_obsolete};
		{set;~obs;
			{regexreplace;
				{get;{get;~leaderboard}};
				/(?:\d*#)(\d{1,18})/g;$1
			}
		}
	}
	{//;Gets the userids based on the page number. If it is page 3, it will set ~1 - ~10 from {get;~0;0} to {get;~0;9}}
	{set;~in1;0}
	{set;~in2;-1}
	{repeat;
		{set;~{increment;~in1};{get;~0;{increment;~in2}}};
		{get;~pgr}
	}
}💴 **| {get;~name} Leaderboard**
```py
📋 Rank | Name

{set;~index;0}
{trim;
	{repeat;
		{if;!=;0;{get;~{increment;~index}};
			[{realpad;{math;+;{get;~index};{get;~mg}};{get;~pad};0;left}]{space;6}>{if;==;{username;{get;~{get;~index}};quiet};{userid;{get;~{get;~index}};quiet};
				#Missing user! - ({get;~{get;~index}}){newline}{space;12}{if;==;1;{get;~del_obsolete};Removed index {indexof;{get;~obs};{get;~{get;~index}}}{void;{splice;{get;{get;~leaderboard}};{indexof;{get;~obs};{get;~{get;~index}}};1}{set;~obs;{regexreplace;{get;{get;~leaderboard}};/(?:\d*#)(\d{1,18})/g;$1}}}};
				#{username;{get;~{get;~index}}}{newline}{space;12}{get;~name}: {get;@{userid;{get;~{get;~index}}}{get;~credit_name}}
			}{newline}
		};
		{get;~pgr}
	}
}

-------------------------------------
# Your {get;~name}: {get;@{userid}{get;~credit_name}}
😐 Rank: {increment;~rank}/{get;~mx}```

File: blargbot/leaderboard/ldmodule.js
{if;==;help;{lower;{args;0}};
**__Leaderboards System Module:__**
**1.** You must run this tag in an `{lb}exec{rb}`
**2.** The arguments used are: ```js
{lb}set{semi}var_name{semi}`name`{rb}
{lb}set{semi}leaderboard_name{semi}`name`{rb}
{lb}set{semi}id{semi}1|0{rb}
{lb}set{semi}crlength{semi}21{rb}

Optional:
{lb}set{semi}rand_user{semi}1|0{rb}
{lb}set{semi}curr_name{semi}`name`{rb}
{lb}set{semi}header{semi}📋 Rank | Name{rb}
{lb}set{semi}show_rank{semi}1|0{rb}

{lb}exec{semi}ldmodule{semi}[page_no]{rb}```
**3.** The `var_name` is the variable of the currency, for example `@{lb}userid{rb}credits`
**4.** The `leaderderboard_name` is the name of the leaderboard variable that will be used.
**5.** The `-id` flag specifies if the user ID is included in the leaderboards.
**6.** The page_no is optional, it defaults to 1.
**7.** You must do `{lb}set{semi}crlength{semi}21{rb}` or the number in which your variable length goes. If it's `5` and user has `600` credits, it becomes `00600`. If it's not specified, it turns to 21 by default.
**8.** `curr_name` is the name of the currency. Credits is the default name.
	{return}
}
{if;!=;;{get;@execldmodule};
	:x: Please run this tag in an `{lb}exec{rb}`. Do `{prefix}t ldmodule help` for more help.
            {return}
}
{if;==;;{get;var_name};
	:x: You didn't provide a variable! Do `{prefix}t ldmodule help` for more help.
		{return}
}
{if;==;;{get;leaderboard_name};
	:x: You didn't provide a leaderboard name! Do `{prefix}t ldmodule help` for more help.
		{return}
}
{if;!=;1;{abs;{floor;{parseint;{get;id}}}};
	{set;id;0}
}
{if;!=;0;{abs;{floor;{parseint;{get;show_rank}}}};
	{set;show_rank;1}
}
{if;==;NaN;{parseint;{get;crlength}};
	{set;crlength;21}
}
{if;<=;{get;crlength};6;
	{if;{get;crlock};;
		:x: Are you sure to set `crlength` to be low? Add `{set;crlock;true}` in your code if you are. 
			{return}
	}
}
{if;==;;{get;curr_name};
	{set;curr_name;Credits}
}
{if;==;1;{get;rand_user};
	{set;~user;{randuser}};
	{set;~user;{userid}}
}
{set;{get;~user}credits;
	{get;{get;var_name}}
}
{if;{isarray;{get;{get;leaderboard_name}}};;
	{set;{get;leaderboard_name};[]}
}
{if;!=;{get;o{get;var_name}};{get;{get;var_name}};
	{set;o{get;var_name};{get;{get;var_name}}}
	{set;@0;
		{regexreplace;
			{get;{get;leaderboard_name}};
			/(\d+#)(\d{1,18})/g;
			$2
		}
	}
	{if;>;
		{length;
			{get;{get;~user}credits}
		};
		{get;crlength};
			{set;crlength;
				{length;
					{get;{get;~user}credits}
				}
			}
	}
	{set;nw;
		{pad;
			left;
			{repeat;0;{get;crlength}};
			0{get;{get;~user}credits}
		}
	}
	{if;==;-1;
		{indexof;
			{get;@0};
			{get;~user}
		};
			{push;{get;leaderboard_name};
				{get;nw}#{get;~user}
			};
			{inject;
				{lb}set{semi}
					{lb}get{semi}leaderboard_name{rb}{semi}
					{lb}regexreplace{semi}
						{lb}get{semi}{lb}get{semi}leaderboard_name{rb}{rb}{semi}
						/\d+(?=#{get;~user})/g{semi}
						{get;nw}
					{rb}
				{rb}
			}
	}
	{sort;{get;{get;leaderboard_name}};descending}
	{set;@0;
		{regexreplace;
			{get;{get;leaderboard_name}};
			/(\d+#)(\d{1,18})/g;
			$2
		}
	}
	{set;~mx;
		{length;{get;@0}}
	}
	{set;~maxpg;
		{ceil;
			{math;/;{get;~mx};10}
		}
	}
	{set;~pg;
		{if;==;0;{argslength};1;
		{if;==;NaN;{parseint;{args;0}};1;
		{if;>;{args;0};{get;~maxpg};{get;~maxpg};
		{if;==;0;{floor;{args;0}};1;
		{abs;{floor;{args;0}}}
	}}}}}
	{set;~mg;
		{math;*;10;
			{math;-;{get;~pg};1}
		}
	}
	{set;@0;
		{slice;
			{regexreplace;
				{get;{get;leaderboard_name}};
				/\d+#(\d{1,18})/g;
				$1
			};
			{get;~mg};
			{math;+;10;{get;~mg}}
		}
	}
	{set;@1;{get;@0;0}}
	{set;@2;{get;@0;1}}
	{set;@3;{get;@0;2}}
	{set;@4;{get;@0;3}}
	{set;@5;{get;@0;4}}
	{set;@6;{get;@0;5}}
	{set;@7;{get;@0;6}}
	{set;@8;{get;@0;7}}
	{set;@9;{get;@0;8}}
	{set;@10;{get;@0;9}}
}```py
{get;header}                  

{trim;{if;!=;NaN;{parseint;{get;@1}};{if;<=;17;{length;{get;@1}};{pad;right;{space;9};[{math;+;1;{get;~mg}}]}> #{username;{get;@1}} {if;==;1;{get;id};({get;@1})}
{space;2}{get;curr_name}: {get;{get;@1}credits}}}
{if;!=;NaN;{parseint;{get;@2}};{if;<=;17;{length;{get;@2}};{pad;right;{space;9};[{math;+;2;{get;~mg}}]}> #{username;{get;@2}} {if;==;1;{get;id};({get;@2})}
{space;2}{get;curr_name}: {get;{get;@2}credits}}}
{if;!=;NaN;{parseint;{get;@3}};{if;<=;17;{length;{get;@3}};{pad;right;{space;9};[{math;+;3;{get;~mg}}]}> #{username;{get;@3}} {if;==;1;{get;id};({get;@3})}
{space;2}{get;curr_name}: {get;{get;@3}credits}}}                     
{if;!=;NaN;{parseint;{get;@4}};{if;<=;17;{length;{get;@4}};{pad;right;{space;9};[{math;+;4;{get;~mg}}]}> #{username;{get;@4}} {if;==;1;{get;id};({get;@4})}
{space;2}{get;curr_name}: {get;{get;@4}credits}}}
{if;!=;NaN;{parseint;{get;@5}};{if;<=;17;{length;{get;@5}};{pad;right;{space;9};[{math;+;5;{get;~mg}}]}> #{username;{get;@5}} {if;==;1;{get;id};({get;@5})}
{space;2}{get;curr_name}: {get;{get;@5}credits}}}
{if;!=;NaN;{parseint;{get;@6}};{if;<=;17;{length;{get;@6}};{pad;right;{space;9};[{math;+;6;{get;~mg}}]}> #{username;{get;@6}} {if;==;1;{get;id};({get;@6})}
{space;2}{get;curr_name}: {get;{get;@6}credits}}}
{if;!=;NaN;{parseint;{get;@7}};{if;<=;17;{length;{get;@7}};{pad;right;{space;9};[{math;+;7;{get;~mg}}]}> #{username;{get;@7}} {if;==;1;{get;id};({get;@7})}
{space;2}{get;curr_name}: {get;{get;@7}credits}}}
{if;!=;NaN;{parseint;{get;@8}};{if;<=;17;{length;{get;@8}};{pad;right;{space;9};[{math;+;8;{get;~mg}}]}> #{username;{get;@8}} {if;==;1;{get;id};({get;@8})}
{space;2}{get;curr_name}: {get;{get;@8}credits}}}                      
{if;!=;NaN;{parseint;{get;@9}};{if;<=;17;{length;{get;@9}};{pad;right;{space;9};[{math;+;9;{get;~mg}}]}> #{username;{get;@9}} {if;==;1;{get;id};({get;@9})}
{space;2}{get;curr_name}: {get;{get;@9}credits}}}
{if;!=;NaN;{parseint;{get;@10}};{if;<=;17;{length;{get;@10}};{pad;right;{space;9};[{math;+;10;{get;~mg}}]}> #{username;{get;@10}} {if;==;1;{get;id};({get;@10})}
{space;2}{get;curr_name}: {get;{get;@10}credits}}}}
                        
-------------------------------------
# Your {get;curr_name}: {get;{get;~user}credits}
{if;==;1;{get;show_rank};😐 Rank: {math;+;1;{indexof;{get;@0};{get;~user}}}/{get;~mx}}```{set;var_name}{set;leaderboard_name}{set;id;0}{set;crlength;21}{set;curr_name;Credits}{set;crlock;false}{set;header;📋 Rank | Name}{set;show_rank;1}{set;rand_user;0}